internal/k8sconfig/leaderelection.go

```diff
@@ -0,0 +1,110 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package k8sconfig // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/k8sconfig"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+// LeaderElectionConfig is used to enable leader election
+type LeaderElectionConfig struct {
+	Enabled bool `mapstructure:"enabled"`
+	// LockName determines the name of the resource that leader election will use for holding the leader lock.
+	LockName      string        `mapstructure:"lock_name"`
+	LeaseDuration time.Duration `mapstructure:"lease_duration"`
+	RenewDeadline time.Duration `mapstructure:"renew_deadline"`
+	RetryPeriod   time.Duration `mapstructure:"retry_period"`
+}
+
+const (
+	inClusterNamespacePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
+	defaultLeaseDuration   = 15 * time.Second
+	defaultRenewDeadline   = 10 * time.Second
+	defaultRetryPeriod     = 2 * time.Second
+)
+
+// NewResourceLock creates a new leases resource lock for use in a leader election loop
+func newResourceLock(client kubernetes.Interface, lockName string) (resourcelock.Interface, error) {
+	if lockName == "" {
+		return nil, errors.New("lockName must be configured")
```

Comment:
From my understanding , Multiple pods preempt the same resource, so there must specify the name of the resource to be preempted, and  during a pod attempts to preempt a resource lock, it carries its own unique identifier. We used the pod uid here. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c70e0043f9145a469a4881a94fd3e5354a4316b6/internal/k8sconfig/leaderelection.go#L47
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c70e0043f9145a469a4881a94fd3e5354a4316b6/internal/k8sconfig/leaderelection.go#L57

Reply:
The lockName shouldn't be just the UID of the controller (deployment, sts, etc.) cause we can have multiple components, e.g. k8sobjectreceiver and kubernetes-cluster receiver, running in the same deployment of collector and trying to elect a leader separately. So we need the lock name to indicate which component it is for. Ideally, the lock name would be unique per component per collector installation.

The `Identity` mentioned above is the id to use to identify the lock holder. Using the pod name (hostname lookup from inside the pod will give the pod name, and not the uid), should be good enough for the id.

Reply:
yes, that's why I use component id as `lock name` here.

Reply:
@jinja2 that makes sense, but we can compose the name from UID of the controller and the component name, right?

Reply:
yes，BTW，could you tell me how to get the UID of collector？you meaning pod uid of collector？

Reply:
It is best to have an UID which multiple otel col shared used.

Reply:
Yes, we can add the uid. Easiest way to get it would be from the owner reference section of the pod, `metadata.ownerReferences[].uid`, but this will be an additional lookup. I don't feel too strongly about adding the uid to the lock name cause how likely are users to run multiple collector deployments with same component ids, and even if they do, they have the option to set the lock name to be non-colliding.

@dmitryax There is no concept of identity in the collector, right? It would be helpful to have a way for collector components to identify themself as part of a specific installation, so you can startup collectors with an id like `--class my-release` and use that for issues like this, instead of relying on the id of platform specific controllers.

Reply:
> It would be helpful to have a way for collector components to identify themself as part of a specific installation, so you can startup collectors with an id like `--class my-release` and use that for issues like this, instead of relying on the id of platform specific controllers.

👍 Good idea. Do I need to create an issue to track it as an discussion?

Reply:
> Yes, we can add the uid. Easiest way to get it would be from the owner reference section of the pod, metadata.ownerReferences[].uid, but this will be an additional lookup

It'll be only one lookup on the start, so should not be a problem

> @dmitryax There is no concept of identity in the collector, right? It would be helpful to have a way for collector components to identify themself as part of a specific installation, so you can startup collectors with an id like --class my-release and use that for issues like this, instead of relying on the id of platform specific controllers.

There is no. Collector always rely on configuration yaml which is mostly decoupled per component. I believe this option (lock_name) should serve that purpose explicitly, but it'd be nice not to make it required for users.

Reply:
@JaredTan95 can you please create an issue to follow up on this?

Reply:
> @JaredTan95 can you please create an issue to follow up on this?

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26367